File: real-time-data-streaming/APIs/nodejs-service/kafkaConsumer.js
const kafka = require('kafka-node');
const { MongoClient } = require('mongodb');

const mongoUrl = 'mongodb://mongo:27017';
const dbName = 'mydatabase';
const collectionName = 'generatedRecords';

async function consumeMessages() {
  const client = new kafka.KafkaClient({ kafkaHost: 'kafka:9092' });
  const consumer = new kafka.Consumer(
    client,
    [{ topic: 'nodejs', partition: 0 }],
    { autoCommit: true }
  );

  const mongoClient = new MongoClient(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true });

  try {
    await mongoClient.connect();
    console.log("Connected to MongoDB");

    const db = mongoClient.db(dbName);
    const collection = db.collection(collectionName);

    consumer.on('message', async function (message) {
      const record = JSON.parse(message.value);
      await collection.insertOne(record);
      console.log("Record inserted into MongoDB", record);
    });
  } catch (err) {
    console.error(err);
  }
}

module.exports = { consumeMessages };

File: real-time-data-streaming/APIs/rails-service/kafkaConsumer.rb
require 'ruby-kafka'
require 'mongo'

mongo_client = Mongo::Client.new(['mongo:27017'], database: 'mydatabase')
collection = mongo_client[:generatedRecords]

kafka = Kafka.new(seed_brokers: ["kafka:9092"], client_id: "rails-service")

consumer = kafka.consumer(group_id: "rails-group")
consumer.subscribe("rails")

trap("TERM") { consumer.stop }

consumer.each_message do |message|
  record = JSON.parse(message.value)
  collection.insert_one(record)
  puts "Record inserted into MongoDB: #{record}"
end
